src/Compilers/Core/Portable/Operations/BasicBlock.cs

```diff
@@ -1,121 +1,113 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
-using System;
-using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Diagnostics;
-using System.Linq;
-using Microsoft.CodeAnalysis.PooledObjects;
 
 namespace Microsoft.CodeAnalysis.FlowAnalysis
 {
-    internal partial class ControlFlowGraphBuilder
+    /// <summary>
+    /// PROTOTYPE(dataflow): Add documentation
+    /// </summary>
+    public sealed class BasicBlock
     {
-        /// <summary>
-        /// PROTOTYPE(dataflow): Rename this type to "BasicBlockBuilder" and source file to "ControlFlowGraphBuilder.BasicBlockBuilder.cs"
-        /// </summary>
-        internal sealed class BasicBlock
+#if DEBUG
+        private bool _sealed;
+#endif
+
+        private ControlFlowBranch _lazySuccessor;
+        private ControlFlowBranch _lazyConditionalSuccessor;
+        private ImmutableArray<ControlFlowBranch> _lazyPredecessors;
+
+        internal BasicBlock(
+            BasicBlockKind kind,
+            ImmutableArray<IOperation> operations,
+            IOperation condition,
+            IOperation value,
+            ControlFlowConditionKind conditionKind,
+            int ordinal,
+            bool isReachable,
+            ControlFlowRegion region)
         {
-            private readonly ArrayBuilder<IOperation> _statementsBuilder;
-            private readonly HashSet<BasicBlock> _predecessorsBuilder;
-            public (IOperation Condition, bool JumpIfTrue, Branch Branch) InternalConditional;
-            public (IOperation Value, Branch Branch) InternalNext;
-            
-            public BasicBlock(BasicBlockKind kind)
-            {
-                Kind = kind;
-                _statementsBuilder = ArrayBuilder<IOperation>.GetInstance();
-                _predecessorsBuilder = new HashSet<BasicBlock>();
-            }
+            Kind = kind;
+            Operations = operations;
+            Condition = condition;
+            Value = value;
+            ConditionKind = conditionKind;
+            Ordinal = ordinal;
+            IsReachable = isReachable;
+            EnclosingRegion = region;
+        }
 
-            public IReadOnlyCollection<IOperation> Statements => _statementsBuilder;
-            public ImmutableArray<IOperation> GetImmutableStatements() => _statementsBuilder.ToImmutableAndFree();
-            public ISet<BasicBlock> Predecessors => _predecessorsBuilder;
-            public BasicBlockKind Kind { get; }
+        public BasicBlockKind Kind { get; }
 
-            public int Ordinal { get; set; } = -1;
+        public ImmutableArray<IOperation> Operations { get; }
```

Comment:
Nit: consider moving the auto-properties up (together with the fields) so it's easier to understand what the block stores.